### PR TITLE
Improve smart contract invocation optimisation.

### PR DESCRIPTION
**Pull request by a-urth**
- added metrics
- restructured code
- benchmark tool changed to spam perl clicker smart contract invocation (to run it requires to spawn it first)
- optimized saving contract memory page
- added optional parameter for benchmark tool `payload` which allows to spam custom transaction, including smart contract
- merged @losfair two PRs implementing caching for smart contracts

### CODE REVIEW

**Comment by a-urth** [comment]
#161

File: tx_applier.go [hunk 1]
```diff
@@ -342,12 +342,12 @@ func executeContractInTransactionContext(
 		}
 		state.GasLimit -= executor.Gas
 
-		logger.Info().
-			Hex("sender_id", tx.Creator[:]).
-			Hex("contract_id", contractID[:]).
-			Uint64("gas", executor.Gas).
-			Uint64("gas_limit", realGasLimit).
-			Msg("Deducted PERLs for invoking smart contract function.")
+		//logger.Info().
```
Comment (a-urth) [inline tx_applier.go]: this logging causes a lot of contention at least while benchmarking locally

File: contract.go [hunk 2]
```diff
@@ -44,6 +45,8 @@ var (
 
 	_ exec.ImportResolver = (*ContractExecutor)(nil)
 	_ compiler.GasPolicy  = (*ContractExecutor)(nil)
+
+	zeroPage = make([]byte, PageSize)
```
Comment (iwasaki-kenta) [inline contract.go]: Care to add the zero page optimization into another PR?

File: contract.go [hunk 3]
```diff
@@ -444,42 +447,25 @@ func LoadContractMemorySnapshot(snapshot *avl.Tree, id AccountID) []byte {
...
-			if len(old) == 0 && mem[pageIdx*PageSize+idx] != 0 {
-				identical = false
-				break
-			}
-
-			if len(old) != 0 && mem[pageIdx*PageSize+idx] != old[idx] {
-				identical = false
-				break
-			}
+		identical, allZero := true, false
```
Comment (iwasaki-kenta) [inline contract.go]: Care to declare `identical, allZero` outside of the for loop, and rename `allZero` to `empty`?

File: contract.go (hunk 2 above)
Comment (iwasaki-kenta) [reply contract.go]: Also, care to declare `zeroPage` into `common.go` instead?

**Comment by iwasaki-kenta** [comment]
Closing as this PR has been split into two separate PRs.

File: contract.go (hunk 2 above)
Comment (a-urth) [reply contract.go]: since this pr is closed i'll move all this whole optimization to separate pr

File: contract.go (hunk 3 above)
Comment (a-urth) [reply contract.go]: yup

**Pull request closed without merging**